### ISSUE TITLE
Migrating the tests

### ISSUE THREAD

**Issue by Rufflewind** [haskell/directory#19]
Currently, the majority of the tests require the framework that's bundled with the GHC repository, which makes testing *directory* independently rather difficult.  All such tests are located under `tests`.

@snoyberg has recently started a stub for tests that can be run under Cabal without the GHC test framework. It could use some love right now.

From here, there are two options I can think of:

  - **Option A**: Port all the tests from GHC framework to the Cabal framework.  Figure out a way to integrate the Cabal tests into GHC's test suite (somehow?).
  - **Option B**: Keep all the tests in the GHC framework.  Figure out a way to run these tests without GHC.

I think Option A is more elegant and less difficult than Option B, since in B one would have to embed/install a minimal GHC framework (somehow).

What might be lost in Option A is the fine-grainedness of the tests: from GHC's perspective it will probably be an "all-or-nothing" test.

Questions:

  - Is there any reason for sticking to GHC's test framework?
  - Any strong feelings about which option is better? (I'm biased to A)
  - Is there's a better idea?

**Comment by ezyang** [comment]
To be quite honest, I don't think Cabal's test framework is very good, and I far prefer working with GHC's test suite which is fast (no need to compile/link the test binaries), easy to write tests for (Cabal tests need more boilerplate, as opposed to the expected stdout/stderr which is very convenient), parallel by default, more flexible (I can write an arbitrary shell script for the test if need be) and generally more convenient to work with. So I'd be more keen on B.

**Comment by Rufflewind** [comment]
Is there an easy way to get a minimal GHC test runner that is not tied specifically to a given GHC repository?  Being able to run it on Travis without downloading and building GHC would be nice.

**Comment by ezyang** [comment]
Well, the test runner is all in Python and can be configured to point to your system GHC, so while the fact that the runner lives in the GHC repo proper might be a pain, you certainly don't have to build GHC.

**Comment by argiopetech** [comment]
I just pulled the git repo (from github, since I couldn't get the canonical link to respect `--depth`) and positioned the `testsuite/` in a suitable location for `directory/test/Makefile` to find it.

```
$ git clone --depth 1 git@github.com:ghc/ghc.git
$ mv ghc/testsuite/ .
$ cd src/directory/tests/
$ make
```

Some time later,

```
Unexpected results from:
TEST="createDirectoryIfMissing001"

OVERALL SUMMARY for test run started at Sat Feb 21 18:40:51 2015 EST
 0:01:31 spent to go through
      13 total tests, which gave rise to
     130 test cases, of which
       2 were skipped

       0 had missing libraries
     118 expected passes
       0 expected failures

       0 caused framework failures
       0 unexpected passes
      10 unexpected failures
       0 unexpected stat failures

Unexpected failures:
   .  createDirectoryIfMissing001 [bad stdout] (normal,hpc,optasm,profasm,threaded1,threaded2,dyn,profthreaded,optllvm)
   .  createDirectoryIfMissing001 [bad stdout or stderr] (ghci)
```

**Comment by Rufflewind** [comment]
Might be related to [this](https://mail.haskell.org/pipermail/ghc-devs/2015-January/008091.html).  I'm fairly certain a large number of the tests in *directory* are not safe to parallelize.

**Comment by ezyang** [comment]
Well, that's well worth fixing!

**Comment by argiopetech** [comment]
Truthfully, I had only intended the above as a demonstration of not needing to build GHC to run the testsuite. The issue it raises does seem to be worth looking into, however.

**Comment by argiopetech** [comment]
Failures in the above were due to accidentally linking against directory-1.2.1.0, so it appears that the technique is fundamentally sound (and the code isn't broken). Assuming `directory` to be in a sandbox, the fix for linking properly is to either run make from a shell spawned with `cabal exec ${SHELL}` or to `cabal sandbox init --sandbox="../.cabal-sandbox"` from `tests/`.

Using the second method,
```
$ cabal exec make
<cut>
OVERALL SUMMARY for test run started at Sun Feb 22 13:33:53 2015 EST
 0:01:41 spent to go through
      13 total tests, which gave rise to
     130 test cases, of which
       2 were skipped

       0 had missing libraries
     128 expected passes
       0 expected failures

       0 caused framework failures
       0 unexpected passes
       0 unexpected failures
       0 unexpected stat failures
```

**Comment by argiopetech** [comment]
As another data point, I've been unable to duplicate this success on Windows. Still looking into why that is (likely PEBCAK), but I can say it isn't as straight forward as running on Unix.

**Comment by Rufflewind** [comment]
I went with Option B since that seems to be the consensus.  The changes can be seen in [23b416 to e85686][1].  Travis CI can now run the entire test suite automatically.  It should also work on Windows with MSYS+Git+Python installed.

In case anyone wants to use them in other packages like [unix][5] or [win32][6]: it can be done with minimal effort, probably without even tweaking the scripts if the relative file structure remains the same.  The crucial files are:

  - [tools/run-tests][2], which downloads the GHC repository and runs the tests;
  - [tools/ghc.patch][3], which fixes `timeout` so it compiles on Windows with older versions of GHC;
  - [tools/dispatch-tests.hs][4], which allows Cabal to invoke the tests.

[1]: https://github.com/haskell/directory/compare/23b416fdca01737223395f1cd243e67e31293101...e856866d6a46a910c3b297ca8c5e6198860f9fec
[2]: https://github.com/haskell/directory/blob/master/tools/run-tests
[3]: https://github.com/haskell/directory/blob/master/tools/ghc.patch
[4]: https://github.com/haskell/directory/blob/master/tools/dispatch-tests.hs
[5]: https://github.com/haskell/unix
[6]: https://github.com/haskell/Win32

**Comment by Rufflewind** [comment]
Reopening this because, after having used it many times now, I really don't think it's worth the trouble anymore:

  - It [interacts poorly with sandboxes](https://github.com/haskell/directory/issues/25) and there's no way to fix that because Cabal does not give us the necessary environment.
  - It forces us to use a Custom `.cabal` to reconcile with Nix's strict dependency management, which
  - by itself was already a bad sign: it means we are bringing in a lot of needless dependencies (Python and `which`, but also many POSIX utilities like `make`).
  - It's very slow on Windows because every test is a separate shell invocation.

Just today I ran into *another* problem where it fails on Windows Command Prompt, because the [GHC test suite changes the codepage to 65001 (UTF-8)](https://github.com/ghc/ghc/blob/9bf003474a377d1f082b58d9366cc595fec01df5/testsuite/driver/runtests.py#L110-L113), which Python 2.7 fails to recognize and immediately bails.  It works fine on MinTTY, and also fine on Python 3+ (probably because [this bug](http://bugs.python.org/issue6058) got fixed).  It's odd that I never discovered this until now, probably because I rarely use Command Prompt.

In summary, it's a very fragile solution: the pieces just don't fit well in Cabal build system and I had to invent several dirty hacks to make it work.

---

Future (long-term) plan: Rewrite a small pure-Haskell test library that mimics the essential features and then port all the tests over.

**Comment by Rufflewind** [comment]
Fixed: https://github.com/haskell/directory/compare/6323523cf08f30b2e4d26ca3a732b827bed551b6...3df3097721660ab934b0731bdf40d0fff5a5ba71
